zetaclient/context/chain.go

```diff
@@ -0,0 +1,166 @@
+type Chain struct {
+	id       int64
+	chain    *chains.Chain
+	params   *observer.ChainParams
+	registry *ChainRegistry
+}
+
+var (
+	ErrChainNotFound     = errors.New("chain not found")
+	ErrChainNotSupported = errors.New("chain not supported")
+)
+
+// NewChainRegistry constructs a new ChainRegistry
+func NewChainRegistry() *ChainRegistry {
+	return &ChainRegistry{
+		chains:           make(map[int64]Chain),
+		additionalChains: []chains.Chain{},
+		mu:               sync.Mutex{},
+	}
+}
+
+// Get returns a chain by ID.
+func (cr *ChainRegistry) Get(chainID int64) (Chain, error) {
+	chain, ok := cr.chains[chainID]
+	if !ok {
+		return Chain{}, ErrChainNotFound
+	}
+
+	return chain, nil
+}
+
+// All returns all chains in the registry sorted by chain ID.
+func (cr *ChainRegistry) All() []Chain {
+	items := maps.Values(cr.chains)
+
+	slices.SortFunc(items, func(a, b Chain) bool { return a.id < b.id })
+
+	return items
+}
+
+// Set sets a chain in the registry.
+// A chain must be SUPPORTED and NOT ZetaChain itself; otherwise returns ErrChainNotSupported
+func (cr *ChainRegistry) Set(chainID int64, chain *chains.Chain, params *observer.ChainParams) error {
+	item, err := newChain(cr, chainID, chain, params)
+	if err != nil {
+		return err
+	}
+
+	item.registry = cr
+
+	cr.mu.Lock()
+	defer cr.mu.Unlock()
+
+	cr.chains[item.id] = item
+
+	return nil
+}
+
+// SetAdditionalChains sets additional chains to the registry
+func (cr *ChainRegistry) SetAdditionalChains(chains []chains.Chain) {
+	cr.mu.Lock()
+	defer cr.mu.Unlock()
+
+	cr.additionalChains = chains
+}
+
+// Delete deletes one or more chains from the registry
+func (cr *ChainRegistry) Delete(chainIDs ...int64) {
+	cr.mu.Lock()
+	defer cr.mu.Unlock()
+
+	for _, id := range chainIDs {
+		delete(cr.chains, id)
+	}
+}
+
+// Has checks if the chain is in the registry
+func (cr *ChainRegistry) Has(chainID int64) bool {
+	_, ok := cr.chains[chainID]
+	return ok
+}
+
+// ChainIDs returns a list of chain IDs in the registry
+func (cr *ChainRegistry) ChainIDs() []int64 {
+	cr.mu.Lock()
+	defer cr.mu.Unlock()
+
+	return maps.Keys(cr.chains)
+}
+
+func newChain(cr *ChainRegistry, chainID int64, chain *chains.Chain, params *observer.ChainParams) (Chain, error) {
+	switch {
+	case chainID < 1:
+		return Chain{}, fmt.Errorf("invalid chain id %d", chainID)
+	case chain == nil:
+		return Chain{}, fmt.Errorf("chain is nil")
+	case params == nil:
+		return Chain{}, fmt.Errorf("chain params is nil")
+	case chain.ChainId != chainID:
+		return Chain{}, fmt.Errorf("chain id %d does not match chain.ChainId %d", chainID, chain.ChainId)
+	case params.ChainId != chainID:
+		return Chain{}, fmt.Errorf("chain id %d does not match params.ChainId %d", chainID, params.ChainId)
+	case !params.IsSupported:
+		return Chain{}, ErrChainNotSupported
+	case chains.IsZetaChain(chainID, nil) || !chain.IsExternal:
+		return Chain{}, errors.Wrap(ErrChainNotSupported, "ZetaChain itself cannot be in the registry")
```

Comment:
This can maybe be moved to a validate sub method

Reply:
https://github.com/zeta-chain/node/pull/2568/commits/2d6e3123e996155b088619478a74076bc14af928